hook-consumer/src/consumer.rs

```diff
@@ -0,0 +1,490 @@
+impl From<HttpMethod> for http::Method {
+    fn from(val: HttpMethod) -> Self {
+        match val {
+            HttpMethod::DELETE => http::Method::DELETE,
+            HttpMethod::GET => http::Method::GET,
+            HttpMethod::PATCH => http::Method::PATCH,
+            HttpMethod::POST => http::Method::POST,
+            HttpMethod::PUT => http::Method::PUT,
+        }
+    }
+}
+
+impl From<&HttpMethod> for http::Method {
+    fn from(val: &HttpMethod) -> Self {
+        match val {
+            HttpMethod::DELETE => http::Method::DELETE,
+            HttpMethod::GET => http::Method::GET,
+            HttpMethod::PATCH => http::Method::PATCH,
+            HttpMethod::POST => http::Method::POST,
+            HttpMethod::PUT => http::Method::PUT,
+        }
+    }
+}
+
+/// `JobParameters` required for the `WebhookConsumer` to execute a webhook.
+/// These parameters should match the exported Webhook interface that PostHog plugins.
+/// implement. See: https://github.com/PostHog/plugin-scaffold/blob/main/src/types.ts#L15.
+#[derive(Deserialize, Serialize, Debug, PartialEq, Clone)]
+pub struct WebhookJobParameters {
+    body: String,
+    headers: collections::HashMap<String, String>,
+    method: HttpMethod,
+    url: String,
+}
+
+/// A consumer to poll `PgQueue` and spawn tasks to process webhooks when a job becomes available.
+pub struct WebhookConsumer<'p> {
+    /// An identifier for this consumer. Used to mark jobs we have consumed.
+    name: String,
+    /// The queue we will be dequeuing jobs from.
+    queue: &'p PgQueue,
+    /// The interval for polling the queue.
+    poll_interval: time::Duration,
+    /// A timeout for webhook requests.
+    request_timeout: time::Duration,
+}
+
+impl<'p> WebhookConsumer<'p> {
+    pub fn new(
+        name: &str,
+        queue: &'p PgQueue,
+        poll_interval: time::Duration,
+        request_timeout: time::Duration,
+    ) -> Self {
+        Self {
+            name: name.to_owned(),
+            queue,
+            poll_interval,
+            request_timeout,
+        }
+    }
+
+    /// Wait until a job becomes available in our queue.
+    async fn wait_for_job<'a>(
+        &self,
+    ) -> Result<PgTransactionJob<'a, WebhookJobParameters>, WebhookConsumerError> {
+        loop {
+            if let Some(job) = self.queue.dequeue_tx(&self.name).await? {
+                return Ok(job);
+            } else {
+                task::sleep(self.poll_interval).await;
+            }
+        }
+    }
+
+    /// Run this consumer to continuously process any jobs that become available.
+    pub async fn run(&self) -> Result<(), WebhookConsumerError> {
+        loop {
+            let webhook_job = self.wait_for_job().await?;
+
+            let request_timeout = self.request_timeout; // Required to avoid capturing self in closure.
+            tokio::spawn(async move { process_webhook_job(webhook_job, request_timeout).await });
```

Comment:
Also, since we `spawn`, what happens to `Errors` returned by `process_webhook_job`? I think we need some wrapper to do logging/metrics? I guess we probably want metrics for `Ok()` results too.

Reply:
One more, I wonder how much we'll care about graceful shutdown? (i.e. it would be nice to not send duplicates every time we do an upgrade) By firing off `spawn`s I'm not sure how we track that there is ongoing work we'd like to give a few seconds to complete?

I guess if we had the counter/semaphore thing in my first comment above, we could start a shutdown (not `dequeue` new tasks) and watch for the counter/semaphore to reach 0?

Reply:
> I think we need a count of outstanding requests to keep a cap on this loop. If I'm reading correctly and we start this process/loop up with, say, 10,000 items sitting the table it will try to run all 10,000 tasks in parallel. Well, that's if it's dequeue (non-txn) based.

Not right now, `dequeue` also acquires a connection, if there are no connections on the pool, then we are blocked on `acquire()`, similar to `dequeue_tx`. So, for now, this is not a problem.

However, if we pass a pool instead of a connection, like you suggested in the comment below, we would block **only on the db calls**. E.g. if our pg connection limit is 10, and all of our 10 worker threads reach a db call, the main task won't be able to acquire a connection to query for new jobs. 

Ignoring that for a moment, what's the problem of high concurrency? We are still going to wait for IO, might as well keep firing requests, why do you think we need a cap (count is okay, always happy to track)? Tokio will evenly distribute work across cpu cores, so we should get even usage across the board, and can bump/autoscale the cpu if that's struggling. 

The only problem I can imagine is lack of host-awareness causing us to involuntarily ddos common hosts. I would prefer exploring a more host-aware cap than a cap on the entire loop, as a general cap would be unfair to less common hosts. But happy to hear other issues, I'm may be missing something. 

> If it's dequeue_tx-based it will be capped on PG waiting for connections. If we're deliberately going to lean on that then I think we should document it here.

Yes. `dequeue_tx` has to hold a transaction, which means we have to hold a connection. I can't think there is a way around this. I guess this is the price we pay and it all comes down to PG performance. Happy to document :+1: 

> Also, since we spawn, what happens to Errors returned by process_webhook_job? I think we need some wrapper to do logging/metrics? I guess we probably want metrics for Ok() results too.

Definitely need metrics here. Probably going to copy the setup at #2 (but without a web middleware).

> One more, I wonder how much we'll care about graceful shutdown?
> I guess if we had the counter/semaphore thing in my first comment above, we could start a shutdown (not dequeue new tasks) and watch for the counter/semaphore to reach 0?

I like this idea, although I would put a time limit on the counter to actually shutdown even if we take too long; I think we can tolerate _some_ duplicates as the expectation with webhooks should be of at least once delivery.

Tokio has great support for graceful shutdown handling: https://tokio.rs/tokio/topics/shutdown. But if you agree I would add this in a follow-up PR for size considerations. Happy to make this a blocking requirement of the overall application, as I think it should be present in v1 (just because it's on a separate PR, doesn't mean I'm pushing this to v2).

Reply:
I've added a comment to reference the current connection limit :+1: 

I agree that `dequeue_tx` still has to be fixed, but the way I think it's broken is that it should be allowed to blast through the queue, basically without limits. I think that there are some fixes to do to address our concerns:

* Memory limits: Enforced limits on payload size (which we can even encode in the type system).
* File descriptors/network (indirectly): Limiting the number of connections reqwest::Client hands out and making new requests wait for it (unfortunately, I can't find anything baked in with Client, but we can wrap this in a tokio Semaphore or something).
* CPU: Tokio should already be dealing with this by setting a default worker limit.